sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,895 @@
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check the validity of the data or count.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries).
+    /// 2. `len` is the correct number of entries (≤ MAX_ENTRIES), matching the prefix.
+    /// 3. The data slice contains at least `NUM_ENTRIES_SIZE + len * ENTRY_SIZE` bytes.
+    /// 4. If `T` is `&[u8]`, that borrow rules are upheld.
+    /// 5. Alignment is correct.
+    ///
+    #[inline(always)]
+    pub unsafe fn new_unchecked_slice(data: &'a [u8], len: usize) -> Self {
+        SlotHashes {
+            data: Data::Slice(data),
+            len,
+        }
+    }
+
+    /// Same as `new_unchecked_slice` but keeps the `Ref` so the runtime borrow
+    /// is held for the lifetime of the `SlotHashes` instance.
+    ///
+    /// # Safety
+    ///
+    /// Items 2 and 4 are normally auto-satisfied in Solana program contexts.
+    ///
+    /// 1. `data` must point to a byte slice that represents **valid** SlotHashes
+    ///    contents for exactly `len` entries (i.e. it was previously validated, or
+    ///    the caller otherwise guarantees correctness).
+    /// 2. The memory backing `data` must remain valid for the entire lifetime `'a`
+    ///    of the returned `SlotHashes` value.
+    /// 3. The pointer in `data` must be correctly aligned for `SlotHashEntry` so
+    ///    that later reference casts are sound.
+    /// 4. Because a [`Ref`] is handed in, the caller must ensure the runtime
+    ///    borrow rules are respected (no mutable aliasing etc.) for as long as
+    ///    the returned `SlotHashes` exists.
+    #[inline(always)]
+    pub unsafe fn new_unchecked_ref(data: Ref<'a, [u8]>, len: usize) -> Self {
+        SlotHashes {
+            data: Data::Ref(data),
+            len,
+        }
+    }
+
+    /// Parses the length prefix of a SlotHashes account and validates that the
+    /// slice is large enough for that many entries.  Only used by the *checked*
+    /// construction paths; unchecked helpers are free to skip this work.
+    ///
+    /// Returns the number of entries on success.
+    #[inline(always)]
+    fn parse_and_validate_data(data: &[u8]) -> Result<usize, ProgramError> {
+        // Need at least the 8-byte length prefix.
+        if data.len() < NUM_ENTRIES_SIZE {
+            return Err(ProgramError::AccountDataTooSmall);
+        }
+        let num_entries =
+            unsafe { ptr::read_unaligned(data.as_ptr() as *const u64) }.to_le() as usize;
+
+        // Reject oversized accounts so callers are not
+        // surprised by silently truncated results.
+        if num_entries > MAX_ENTRIES {
+            return Err(ProgramError::InvalidAccountData);
+        }
+
+        let required_len = NUM_ENTRIES_SIZE + num_entries * ENTRY_SIZE;
+        if data.len() < required_len {
+            return Err(ProgramError::InvalidAccountData);
```

Comment:
Maybe we should use `ProgramError::InvalidArgument`, since the data might not come from an account.

Reply:
Makes sense to me here, changed in caf715773a5a2dcac542baf0051f9e4c211e241d (except in case of `from_account_info`